craft_store/store_client.py

```diff
@@ -0,0 +1,169 @@
+
+    def __init__(
+        self,
+        *,
+        base_url: str,
+        endpoints: "endpoints.Endpoints",
+        application_name: str,
+        user_agent: str,
+    ) -> None:
+        """Initialize the Store Client.
+
+        :param base_url: the base url of the API endpoint.
+        :param endpoints: :data:`endpoints.CHARMHUB` or :data:`endpoints.SNAP_STORE`.
+        :param application_name: the name application using this class, used for the keyring.
+        :param user_agent: User-Agent header to use for HTTP(s) requests.
+        """
+        super().__init__(user_agent=user_agent)
+
+        self._bakery_client = httpbakery.Client()
+        self._base_url = base_url
+        self._store_host = urlparse(base_url).netloc
+        self._endpoints = endpoints
+        self._auth = Auth(application_name, base_url)
+
+    def _get_token(self, token_request: Dict[str, Any]) -> str:
+        token_response = super().request(
+            "POST",
+            self._base_url + self._endpoints.tokens,
+            json=token_request,
+        )
+
+        return token_response.json()["macaroon"]
+
+    def _candid_discharge(self, macaroon: str) -> str:
+        bakery_macaroon = bakery.Macaroon.from_dict(json.loads(macaroon))
+        discharges = bakery.discharge_all(
+            bakery_macaroon, self._bakery_client.acquire_discharge
+        )
+
+        # serialize macaroons the bakery-way
+        discharged_macaroons = (
+            "[" + ",".join(map(utils.macaroon_to_json_string, discharges)) + "]"
+        )
+
+        return base64.urlsafe_b64encode(utils.to_bytes(discharged_macaroons)).decode(
+            "ascii"
+        )
+
+    def _authorize_token(self, candid_discharged_macaroon: str) -> str:
+        token_exchange_response = super().request(
+            "POST",
+            self._base_url + self._endpoints.tokens_exchange,
+            headers={"Macaroons": candid_discharged_macaroon},
+        )
+
+        return token_exchange_response.json()["macaroon"]
+
+    def login(
+        self,
+        *,
+        permissions: Sequence[str],
+        description: str,
+        ttl: str,
+    ) -> None:
+        """Obtain credentials to perform authenticated requests.
+
+        Credentials are stored on the systems keyring, handled by
+        :data:`craft_store.auth.Auth`.
+
+        The list of permissions to select from can be referred to on
+        :data:`craft_store.attenuations`.
+
+        :param permissions: Set of permissions to grant the login.
+        :param description: Client description to refer to from the Store.
+        :param ttl: time to live for the credential, in other words, how
+                    long until it expires, expressed in seconds.
+        """
+        token_request = self._endpoints.get_token_request(
+            permissions=permissions, description=description, ttl=ttl
+        )
+
+        token = self._get_token(token_request)
```

Comment:
Maybe call this variable `macaroon` to make its nature more explicit?

Reply:
updated in https://github.com/canonical/craft-store/pull/6/commits/84fa660f874ab10b86dcf59eb1fa68c141c64234